### ISSUE TITLE
matplotlib breaks `HiddenMarkovModel.fit` if the states are modelled as `MultivariateGaussianDistribution`

### ISSUE THREAD

**Issue by paulbrodersen** [jmschrei/pomegranate#508]
I fully realize that the following may sound insane but initialising a figure with `matplotlib` does not seem to play nicely with the `MultivariateGaussianDistribution`.  Specifically, initialising and showing a figure prevents a successful fit when training an HMM ("Improvement: nan"). This does not seem to occur with other distributions (at least not the `NormalDistribution`, which I also tested).  

The following MWE:

```python
#!/usr/bin/env python
import numpy as np
import matplotlib.pyplot as plt
import pomegranate

if __name__ == '__main__':

    ddir = '../data/'
    signals = np.load('test.npy')
    with open('model.json', 'r') as f:
        json_object = f.read()

    model = pomegranate.HiddenMarkovModel.from_json(json_object)

    for ii in range(3):

        model = model.fit(signals,
                          algorithm='baum-welch',
                          inertia=0.95,
                          max_iterations=3,
                          verbose=True,
                          use_pseudocount=True,
        )

        # --------------------------------------------------------------------------------
        # THE OFFENDING LINE
        fig, ax = plt.subplots(1,1)
        # --------------------------------------------------------------------------------

        plt.ion()
        plt.show()
        input('Press any key to close figures...')
        plt.close('all')
```

[link](https://drive.google.com/file/d/10v8EtthpVbUs6aFoWJt9jd_VgN2YErr3/view?usp=sharing) to `test.npy`

[link](https://drive.google.com/open?id=1vD_SHgYLb8nhlV5QoRzBO4wWREB2HFi4) to `model.json`

Fitting multivariate Gaussian distributions seems to be a bit unstable in general, so you may need to rerun the script a few times to get successful fit even on the first iteration. 

matplotlib version: 3.0.0
pomegranate version: 0.10.0
python 3.5.3

**Comment by paulbrodersen** [comment]
I also seem to get the same problem with `IndependentComponentsDistribution`.

**Comment by paulbrodersen** [comment]
I am just glad the bug is reproducible. It was driving me absolutely mad when it happened.

**Comment by kroegern1** [comment]
I'm so sorry you had to endure that, but it saved me a lot of resources haha.
Do we know why training is unstable? Why are there NaNs only sometimes when you run it with Multivariate Gaussian?

**Comment by kroegern1** [comment]
```
retrain = True
while retrain:
    transmat           = getTransitionMatrix(num_states)
    distributions      = getDistrbutions(num_states, dim)
    startprob, endprob = getStartAndEndProb(num_states)

    model = HiddenMarkovModel.from_matrix(transmat, distributions, startprob, endprob)

    model.bake()

    result = "\n".join( "{}: {}".format( state.name, state.distribution.parameters )
                        for state in model.states if not state.is_silent() )
    print(result)

    print("\n\nFITTING MODEL \n\n")
    model.fit(np.array(dataset))  # needs to be list of numpy arrays

    #Check for NaNs in Multivariate gaussian distribution.
    # If there are any NaNs, retrain the model due to stability issues.
    nan_states = []
    for state in model.states:
        if not 'None' in state.name:  # valid state
            mu_learned = state.distribution.parameters[0]
            if np.isnan(mu_learned).any():
                nan_states.append(True)

    if np.array(nan_states).any():  # there are nans, retrain!
        retrain = True
        print("There are NaNs!")
        Model = None #reset the model
    else:
        retrain = False

    result = "\n".join( "{}: {}".format( state.name, state.distribution.parameters )
                        for state in model.states if not state.is_silent() )
    print(result)

```

This is my hack solution of forcing a model to be trained without NaNs .... just keep retraining until it works. I don't like this solution, but it's the best I can think of for now.

**Comment by paulbrodersen** [comment]
> Now that I have this retraining procedure, I added back in the plotting code and I'm able to converge on some trained model, so I don't know if this theory actually holds. 

On my machine, with the (synthetic/simulated) inputs and the (known to be correct) model that I linked, the issue was highly reproducible. Without figure initialisation, I usually get the correct model; with the figure, I never do.

**Comment by kroegern1** [comment]
Update for you Paul and I suppose any others that read this. (I believe your plotting code could also cause a problem, but that was not necessarily the error with my code).

For me, the error in training was from a "nan" improvement on iteration 1 when I have the verbose flag in model.fit() to true. The reason under the hood of why this was happening is not clear to me, but  basically I was setting some of my feature dimensions in each sequence to 0.0001, and when I changed it to 0.0 padding, it worked. So I guess baum-welch didn't like my small value padding and liked 0.0 more.

**Comment by jmschrei** [comment]
Thank you for opening an issue. pomegranate has recently been rewritten from the ground up to use PyTorch instead of Cython (v1.0.0), and so all issues are being closed as they are likely out of date. Please re-open or start a new issue if a related issue is still present in the new codebase.
